### PR TITLE
:sparkles: Port rework

### LINKED ISSUE #1023: Consolidate Networks and Ports in OpenStackMachineSpec
/kind feature

**Describe the solution you'd like**
Currently in MachineSpec we have 2 different ways to add a vif to a VM: [Networks](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/007a8278293efdded741831de584fcc4a1d6867e/api/v1alpha4/openstackmachine_types.go#L55) and [Ports](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/007a8278293efdded741831de584fcc4a1d6867e/api/v1alpha4/openstackmachine_types.go#L59) which are defined as lists of [NetworkParam](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/007a8278293efdded741831de584fcc4a1d6867e/api/v1alpha4/types.go#L57-L68) and [PortOpts](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/007a8278293efdded741831de584fcc4a1d6867e/api/v1alpha4/types.go#L123-L154) respectively.

The difference can be summarised as:

Networks:
* Networks and Subnets can be specified explicitly (which is also efficient since  #994) or by filter.
* Cannot specify port options

Ports:
* Networks and Subnets must be specified explicitly
* Possible specify a port with no subnet
* Can specify all port options

Apart from being confusing for the user, the difference is also a source of potential maintenance bugs and different ports can accidentally end up with slightly different semantics.

I think we could easily create a consolidated API which combines the flexible network selection of Networks with the ability to specify all port options of Ports.

### PR DESCRIPTION
**Pull request by iamemilio**
**What this PR does / why we need it**: Makes the port api powerful enough to replace the network/subnet api.

Fixes #1023

- NOTE: Does not depricate networks or subnets yet

**Special notes for your reviewer**: Needs testing, preferably on cluster with multiple networks/subnets

### CODE REVIEW

File: api/v1alpha4/types.go [hunk 1]
```diff
@@ -121,8 +121,9 @@ type SubnetFilter struct {
 }
 
 type PortOpts struct {
-	// ID of the OpenStack network on which to create the port. If unspecified, create the port on the default cluster network.
-	NetworkID string `json:"networkId,omitempty"`
```
[the remark this answers is not in the document]
Comment (iamemilio) [reply api/v1alpha4/types.go:125]: +1  I dont think the port interface lends itself well to creating more than a single port, and I think allowing more than 1 network to be returned here would be bad UI. I would be on team API bump

File: pkg/cloud/services/compute/instance.go [hunk 2]
```diff
@@ -135,9 +135,18 @@ func (s *Service) constructNetworks(openStackCluster *infrav1.OpenStackCluster,
...
 			pOpts.Trunk = &openStackMachine.Spec.Trunk
 		}
-		if port.NetworkID != "" {
+		if port.Network != nil {
+			netIDs, err := s.networkingService.GetNetworkIDsByFilter(networks.ListOpts(*port.Network))
+			if err != nil {
+				return nil, err
+			}
+			if len(netIDs) > 1 {
+				return nil, fmt.Errorf("network filter for port %s returns more than one result", port.NameSuffix)
```
[the remark this answers is not in the document]
Comment (iamemilio) [reply pkg/cloud/services/compute/instance.go:155]: closing due to discussion that happened here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/1023#issuecomment-957951927

**Comment by iamemilio** [comment]
I will Squash these commit down, they are not neat.

File: pkg/cloud/services/networking/port_test.go [hunk 3]
```diff
@@ -225,12 +229,68 @@ func Test_GetOrCreatePort(t *testing.T) {
...
+					},
+				}, nil)
 			},
 			&ports.Port{
 				ID: portID1,
 			},
 			false,
 		},
+		{
+			"fails to create port with specified portOpts if subnet query returns more than one subnet",
```
Comment (macaptain) [inline pkg/cloud/services/networking/port_test.go:250]: Maybe we want a similar test for if the networks query returns more than one network?

**Comment by jichenjc** [comment]
code looks reasonable, just need take this into v-Next plan (alpha5 or beta1 ?)
also, need follow up to some doc update/template update as well?

/approve

File: pkg/cloud/services/networking/port_test.go (hunk 3 above)
Comment (iamemilio) [reply pkg/cloud/services/networking/port_test.go:250]: I made a new suite to test this. Check the second commit

**Comment by iamemilio** [comment]
IMO, the code here could land in v4 or v5. It should not be that disruptive. The deprecation of the network/subnet api is going to have to be in either v5 or beta1

**Comment by iamemilio** [comment]
I would like to push for this in v1Alpha5

**Comment by jichenjc** [comment]
check this 
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/1024#issuecomment-968406388

I think given we merged https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1041

we should be good to take this in ? @hidekazuna please comment, thanks

**Comment by macaptain** [comment]
Hey @iamemilio, sorry for the delay on this, I finally got around to testing this branch manually on the happy path.

Config:

```
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
kind: OpenStackMachineTemplate
metadata:
  name: basic-1-md-0
  namespace: default
spec:
  template:
    spec:
...
      ports:
      - network:
          name: mcaptain-capo-test
```

And my machine was created with port on my test network specified by name, as expected. The code looks good too, so I am happy to:

/lgtm

I'll also

/hold

in case you wanted to do anything more before merging. But all good from what I can see. Thanks!

**Comment by iamemilio** [comment]
This is ready to ship :ship:

**Comment by hidekazuna** [comment]
@iamemilio Now v1beta1 API is merged so that update to change v1beta1 API, please.

**Comment by iamemilio** [comment]
/hold cancel

lgtm when ready

**Comment by iamemilio** [comment]
/hold 
are the e2e tests still using alpha4? Also, do we have tests for the conversion functions in place? I think this breaks them right now.

**Comment by hidekazuna** [comment]
@iamemilio e2e test uses v1veta1: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/tree/main/test/e2e/data/infrastructure-openstack
But we do not have upgrade test like CAPA: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/test/e2e/data/e2e_conf.yaml#L78-L98

**Comment by iamemilio** [comment]
My hunch is that I introduced a nil pointer error in the v1alpha4 conversion function, and its failing to build the webhook, lets see if this fixes it

**Comment by iamemilio** [comment]
dear god, what have I done

**Pull request closed without merging**